plugins/s3_auth/aws_auth_v4.cc

```diff
@@ -0,0 +1,730 @@
+/*
+  Licensed to the Apache Software Foundation (ASF) under one
+  or more contributor license agreements.  See the NOTICE file
+  distributed with this work for additional information
+  regarding copyright ownership.  The ASF licenses this file
+  to you under the Apache License, Version 2.0 (the
+  "License"); you may not use this file except in compliance
+  with the License.  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+
+/**
+ * @file aws_auth_v4.cc
+ * @brief AWS Auth v4 signing utility.
+ * @see aws_auth_v4.h
+ */
+
+#include <cstring>        /* strlen() */
+#include <ctime>          /* strftime(), time(), gmtime_r() */
+#include <iomanip>        /* std::setw */
+#include <sstream>        /* std::stringstream */
+#include <openssl/sha.h>  /* SHA(), sha256_Update(), SHA256_Final, etc. */
+#include <openssl/hmac.h> /* HMAC() */
+
+#undef AWS_AUTH_V4_DETAILED_DEBUG_OUTPUT
+#ifdef AWS_AUTH_V4_DETAILED_DEBUG_OUTPUT
+#include <iostream>
+#endif
+
+#include "aws_auth_v4.h"
+
+/**
+ * @brief Lower-case Base16 encode a character string (hexadecimal format)
+ *
+ * @see AWS spec: http://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-header-based-auth.html
+ * Base16 RFC4648: https://tools.ietf.org/html/rfc4648#section-8
+ *
+ * @param in ptr to an input counted string to be base16 encoded.
+ * @param inLen input character string length
+ * @return base16 encoded string.
+ */
+String
```

Comment:
Interesting. We have an internal lib/ts/ink_base64.cc, I wonder if it'd be useful to have generic support for all the various Base\<nn\> flavors in the core, and expose those as TS APIs? Not saying this has to be done here, as part of this PR, but maybe it'd be generally useful to have long-term? Similar to how we have

```
TSReturnCode
TSBase64Decode(const char *str, size_t str_len, unsigned char *dst, size_t dst_size, size_t *length)

TSReturnCode
TSBase64Encode(const char *str, size_t str_len, char *dst, size_t dst_size, size_t *length)
```